### ISSUE TITLE
Expose statically typed way to get convention from sourceSet

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/gradle#3191]
### Ask

Java/Kotlin developers writing plugins need a way to get a `SourceSet` convention off of the `SourceSet` object.

### Expected Behavior

There should be a public (non `internal`) API exposed for getting a convention off of a source set.

### Current Behavior
Currently, the only way to retrieve language specific source sets off of the `SourceSet` object is to use one of the following patterns:

Example from the [Gradle Kotlin DSL project](https://github.com/gradle/kotlin-dsl/blob/6160cb88a99f846ad34d916f3847563532c46ac8/buildSrc/src/main/kotlin/build/KotlinPluginExtensions.kt):
```kotlin
import org.gradle.api.internal.HasConvention

val SourceSet.kotlin: SourceDirectorySet
    get() =
        (this as HasConvention)
            .convention
            .getPlugin(KotlinSourceSet::class.java)
            .kotlin
```

Example from the [diffplug/spotless plugin](https://github.com/diffplug/spotless/blob/166195d549ae2a7f4b703717af94943d81f1b8a1/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java#L133):

```java
import org.gradle.api.internal.plugins.DslObject;
// In function
GroovySourceSet groovySourceSet = new DslObject(sourceSet).getConvention().getPlugin(GroovySourceSet.class);
```

### Context
There is no way to get the `GroovySourceSet` or `KotlinSourceSet` off of the `SourceSet` object.

### Your Environment

The `DslObject` comes from a plugin using the Gradle `2.14.1` API.
The kotlin example is using whatever the bleeding edge of gradle is (just assume `4.2.1`)

### Minimal request

I need a work around that allows me to do this without using internal API's:
This is for a plugin that is built using gradle version `2.14.1`.
```
GroovySourceSet groovySourceSet = new DslObject(sourceSet).getConvention().getPlugin(GroovySourceSet.class);
```

**Comment by JLLeitschuh** [comment]
@oehme Is there a way to do this without using reflection using the `2.14.1` API?

Also, could this simply be fixed by exposing the `HasConvention` as a part of the public API?

**Comment by oehme** [comment]
We want to deprecate `Convention`, so spreading it wider is not an option I'd say. I think the proper fix here is to provide a way to model this as an `Extension` instead and deprecate/remove all uses of `Convention` from our code base.

**Comment by JLLeitschuh** [comment]
Can you explain the intention behind deprecating `Convention` (basically, why this is an outdated idea) and explain how modeling this as an `Extension` would work (perhaps with a minimalist API consumer example)??

**Comment by oehme** [comment]
`Convention` has been superseded since the introduction of `Extension`. It's only around because some plugins used it from the start (e.g. the Java plugin) and we didn't get around to cleaning it up. Source sets are another example that really aren't modelled right. The top-level elements should be components and the source sets should be all about languages. E.g.

```
components {
  main(JavaLibrary) {
    sourceSets {
      scala(ScalaSourceSet) {
         //....
      }
    }
  }
}
```

We are slowly working towards this better model, but of course we also need an answer for the existing `SourceSet` abstraction. 

One answer would be to add some backwards compatiblity handling that keeps the Groovy DSL working, while replacing the actual data object with an extension. I.e. you'd then cast the sourceSet to `ExtensionAware` and retrieve the `ScalaSourceSetExtension` from it. We could even auto-generate a static accessor for it.

Another workaround would be for the Kotlin DSL to generate static extension methods for every convention it discovers and doing so recursively through the whole DSL object graph. 

Basically anything that doesn't directly expose `Convention` to the user is cool.

**Comment by JLLeitschuh** [comment]
Just making sure that I understand, `components` is an, as of currently, unreleased/unsupported mechanism? (It doesn't show up as a supported keyword in the Gradle Kotlin DSL on `Project`).
`main` would be accessing code in the `main` component??
`JavaLibrary` would be the `org.gradle.api.internal.java.JavaLibrary` that currently implements `SoftwareComponentInternal`?
`sourceSets` would have to be added to the `SoftwareComponent` type?

**Comment by oehme** [comment]
You nailed it 👍 

The `components {}` block is currently only available in the software model, but we're porting the concept back into the current model.

**Comment by bamboo** [comment]
@JLLeitschuh [v0.11.1](https://github.com/gradle/kotlin-dsl/releases/tag/v0.11.1) introduced the `withConvention` utility for that exact use-case. Here's how it could be used to rewrite that first code sample:

```kotlin
val SourceSet.kotlin: SourceDirectorySet
    get() = withConvention(KotlinSourceSet::class) { kotlin }
```

**Comment by JLLeitschuh** [comment]
That solves the Gradle Kotlin DSL problem ( 🎉  ) but doesn't solve my problem for trying to refactor this [line in diffplug/spotless plugin](https://github.com/diffplug/spotless/blob/166195d549ae2a7f4b703717af94943d81f1b8a1/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java#L133) to not use any internal API's.

**Comment by bamboo** [comment]
> ... doesn't solve my problem for trying to refactor this line in diffplug/spotless plugin to not use any internal API's.

Right.

**Comment by ilya-g** [comment]
> `val SourceSet.kotlin: SourceDirectorySet`

@bamboo Would it be a good idea if we introduce this extension in kotlin gradle plugin [KT-19472](https://youtrack.jetbrains.com/issue/KT-19472)?

**Comment by Mazorius** [comment]
This problem also occurred in a Gradle plugin written with Kotlin 1.3.71.

In the Jenkins Shared Libraries the normal folder structures are not used.

I have now solved this, but I think it should work without an internal method.

```
private fun configureJenkinsSharedLibrarySourceSets(
        project: Project,
        mainSourceSet: SourceSet
    ) {
        project.logger.debug("Apply Jenkins shared library source set configuration")

        project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets.forEach {
            if (it.name == SourceSet.MAIN_SOURCE_SET_NAME) {
                DslObject(it).convention.getPlugin(GroovySourceSet::class.java).groovy.srcDirs("src", "vars")
            }
            if (it.name == SourceSet.TEST_SOURCE_SET_NAME) {
                DslObject(it).convention.getPlugin(GroovySourceSet::class.java).groovy.srcDirs("test")
            }
        }
        mainSourceSet.resources.srcDirs("resources")
    }
```

Maybe someone can give me a better solution.

**Comment by jjohannes** [comment]
We are still missing conveniences in Kotlin DSL if a _source directory set_ was added. You can write you own accessory like this:

```
val SourceSet.groovy: SourceDirectorySet
    get() = withConvention(GroovySourceSet::class) { groovy }


val SourceSet.kotlin: SourceDirectorySet
    get() = withConvention(KotlinSourceSet::class) { kotlin }
```

But these should be generated by Gradle automatically.

**Comment by fcurts** [comment]
@jjohannes Have you used these extension methods for real? Don't work for me.

**Comment by big-guy** [comment]
This is going to be delivered in 7.1 with new Groovy/Scala/Antlr extensions:

```
sourceSets {
    main {
        groovy {
            // configure the Groovy source set
        }
    }
}
```

This isn't exactly the same thing as what was asked here because we're deprecating/getting rid of the convention types eventually.  The extensions are replacements for them.
